### PR TITLE
Enable perioding resuming to fetch logs for KPOA

### PR DESCRIPTION
**Pull request by dstandish**
This PR modifies Async KPO to periodicall resume the sync portion of the task to fetch and emit the latest logs before deferring again.  

To control this behavior a new param is added: `logging_interval`.  If `None` then Async KPO operates as it did before, never resuming the sync task until the pod is done running.  If logging_interval is given a value (integer seconds) then the trigger will resume every `logging_interval` seconds to fetch latest logs and defer again (until pod is done).

Here's a task example you can test this with:

```python
test_kubernetes_pod = KubernetesPodOperatorAsync(
    namespace='airflow',
    image='python:3.8-slim',
    cmds=[
        "bash",
        "-cx",
        (
            "i=0; "
            "while [ $i -ne 30 ]; "
            "do i=$(($i+1)); "
            "echo $i; "
            "sleep 1; "
            "done; "
            "mkdir -p /airflow/xcom/; "
            "echo '{\"message\": \"good afternoon!\"}' > /airflow/xcom/return.json"
        ),
    ],
    name="test_kubernetes_pod",
    in_cluster=False,
    task_id="test_kubernetes_pod",
    logging_interval=10,
    dag=dag,
    is_delete_operator_pod=False,
    # do_xcom_push=True,
)
```

### CODE REVIEW

**Comment by dstandish** [comment]
@kaxil this is ready for review but not for merge because it needs to wait on another k8s provider release
tests passing locally now though

File: astronomer/providers/cncf/kubernetes/triggers/wait_container.py
Comment (kaxil) [inline astronomer/providers/cncf/kubernetes/triggers/wait_container.py:51]: Need to add this to docstrings

**Review by kaxil — COMMENTED** [review]
This looks great!! Have you tested in on Astro Cloud or locally via chart?

**Comment by dstandish** [comment]
> This looks great!! Have you tested in on Astro Cloud or locally via chart?

Only locally... it should be same now that we have gotten that service account issue resolved... but i can work on that if you think it's important to do.

**Comment by kaxil** [comment]
> > This looks great!! Have you tested in on Astro Cloud or locally via chart?
> 
> Only locally... it should be same now that we have gotten that service account issue resolved... but i can work on that if you think it's important to do.

@phanikumv Can you assign this to someone in the team to test too please?

> Related note, are we gonna OSS the istio stuff?

Most-likely yes, in a way that is reusable for the OSS community.

**Comment by phanikumv** [comment]
@pankajastro please test this once merged to main as part of our integration test

**Comment by dstandish** [comment]
this is on hold until next provider release

**Comment by raphaelauv** [comment]
@dstandish apache-airflow-providers-cncf-kubernetes 4.0.0 is out :+1:

**Comment by kaxil** [comment]
```
apache-airflow-providers-cncf-kubernetes 4.0.0 depends on apache-airflow>=2.3.0
```
https://app.circleci.com/pipelines/github/astronomer/astronomer-providers/1165/workflows/dfda1e9e-426b-437e-8825-7c9e4bc4ccce/jobs/3125
😮

**Comment by kaxil** [comment]
https://github.com/apache/airflow/blob/0cd8833df74f4b0498026c4103bab130e1fc1068/airflow/providers/cncf/kubernetes/provider.yaml#L45

I don't think this was intentional, we don't need to depend on Airflow 2.3.0 @dstandish - Let's fix this in OSS

**Comment by raphaelauv** [comment]
@kaxil since astronomer-providers is not related to the KubernetesExecutor

it could be based on 

```
kubernetes==21.7.0
apache-airflow-providers-cncf-kubernetes==4.0.0rc1 --no-deps
```

https://github.com/apache/airflow/issues/22836#issuecomment-1093012855

**Comment by kaxil** [comment]
> @kaxil since astronomer-providers is not related to the KubernetesExecutor
> 
> it could be based on
> 
> ```
> kubernetes==21.7.0
> apache-airflow-providers-cncf-kubernetes==4.0.0 --no-deps
> ```
> 
> [apache/airflow#22836 (comment)](https://github.com/apache/airflow/issues/22836#issuecomment-1093012855)

Even if we do that, all the users of this provider will have to do the same to install it :( 

Let me think more about it and see if it is better to release a new version of Kubernetes Provider on Apache Airflow with just the logging related changes without the k8s client change.

**Comment by kaxil** [comment]
@dstandish Now that 2.3 is out and new provider version is out, can we resurrect this PR?

**Comment by dstandish** [comment]
> @dstandish Now that 2.3 is out and new provider version is out, can we resurrect this PR?

sure just rebased.  not seeing any build information on the pr.

File: astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py
Comment (kaxil) [inline astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py:49]:
```suggestion
        """Defers to ``WaitContainerTrigger`` optionally with last log time."""
```

**Comment by kaxil** [comment]
> > @dstandish Now that 2.3 is out and new provider version is out, can we resurrect this PR?
> 
> sure just rebased. not seeing any build information on the pr.

Pushed a dummy commit -- that should trigger a build. Not sure why it didn't build before 🤷

File: astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py
Comment (kaxil) [inline astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py:32]: Can we add this to docstring above on L29 and update the warning on L25 too please.

And looks like tests are passing 🎉 

cc @pankajastro / @pankajkoti  / @phanikumv can one of you test this on our deployment to see the logging behaviour?

Comment (pankajkoti) [reply astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py:32]: Looks fine to me. I have set the interval to 10 seconds, but across 3 separate runs it resumed to gather the logs with an interval of 15-17 seconds and then got deferred again.

Comment (pankajkoti) [reply astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py:32]: https://user-images.githubusercontent.com/10206082/168273415-f5194af5-0bcd-4959-8e14-1ed62fb32f1c.mp4

@dstandish Attaching a screen-recording for the observed run

Comment (pankajkoti) [reply astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py:32]: Also, @kaxil  has the following points:

1. Should we disable the mini scheduler once we go-back to the worker just for fetching logs?
2. We probably want to add a log line saying - "Task is not completed, we are just fetching logs periodically" etc

Comment (dstandish) [reply astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py:32]: > Can we add this to docstring above on L29 and update the warning on L25 too please.

done

Comment (dstandish) [reply astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py:32]: > * Should we disable the mini scheduler once we go-back to the worker just for fetching logs?
> * We probably want to add a log line saying - "Task is not completed, we are just fetching logs periodically" etc

not familiar with the mini scheduler. @kaxil can you point me to any info / docs / code?  

re logging, one thing though, when we resume, we don't necessarily know that the task isn't done.  it could have finished after trigger fired "resume" event but before task is picked up.  or it could finish during logs read.  we get the container status immediately after fetching the latest logs, and at that time if it's still running we defer again.  but i suppose we could add a line that explains why we're deferring again...

Comment (dstandish) [reply astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py:35]: > but i suppose we could add a line that explains why we're deferring again..

done this now

Comment (kaxil) [reply astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py:35]: @dstandish Mini-scheduler is the scheduling that happens on workers for optimizations controlled by [[scheduler] schedule_after_task_execution](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#schedule-after-task-execution)

Comment (kaxil) [reply astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py:35]: and also mentioned in https://airflow.apache.org/docs/apache-airflow/stable/concepts/scheduler.html?highlight=schedule_after_task_execution#scheduler-configuration-options

**Comment by kaxil** [comment]
Can you try the following @dstandish  -- to see why CircleCI isn't creating the job for your user .. 

https://support.circleci.com/hc/en-us/articles/360044395953-GitHub-Updates-Results-in-400-Response-for-Webhook

File: astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py
Comment (kaxil) [inline astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py:26]:
```suggestion
        you can configure ``KubernetesPodOperatorAsync`` to periodically resume and fetch logs.  This behavior
```

File: astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py
Comment (kaxil) [reply astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py:35]: One Mypy failure too::

```
astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py: note: In member "defer" of class "KubernetesPodOperatorAsync":
astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py:51: error:
Function is missing a type annotation  [no-untyped-def]
        def defer(self, last_log_time=None, **kwargs):
        ^
Found 1 error in 1 file (checked 136 source files)

Exited with code exit status 1

```

File: astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py
Comment (kaxil) [inline astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py:32]:
```suggestion
        resuming to fetch latest logs. If ``None``, then the task will remain in deferred state until pod
```

**Comment by kaxil** [comment]
@dstandish Mypy still complaining :( 

```
stronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py: note: In member "defer" of class "KubernetesPodOperatorAsync":
astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py:52: error:
Function is missing a type annotation for one or more arguments 
[no-untyped-def]
        def defer(self, last_log_time: Optional[DateTime] = None, **kwargs...
        ^
astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py:67: error:
Item "None" of "Optional[Any]" has no attribute "metadata"  [union-attr]
                    pod_name=self.pod.metadata.name,
                             ^
astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py:69: error:
Item "None" of "Optional[Any]" has no attribute "metadata"  [union-attr]
                    pod_namespace=self.pod.metadata.namespace,
                                  ^
astronomer/providers/cncf/kubernetes/operators/kubernetes_pod.py:73: error:
Argument "last_log_time" to "WaitContainerTrigger" has incompatible type
"Optional[DateTime]"; expected "Optional[int]"  [arg-type]
                    last_log_time=last_log_time,
                                  ^
Found 4 errors in 1 file (checked 136 source files)

Exited with code exit status 1

```

**Comment by phanikumv** [comment]
@pankajastro will fix the mypy issue and get this merged to main

**Comment by pankajastro** [comment]
> 8.28%



> @pankajastro will fix the mypy issue and get this merged to main

The unit test coverage are passing now.

**Review by phanikumv — APPROVED** [review]
Good job @pankajastro on getting the mypy fixes done quickly.

**Comment by pankajastro** [comment]
Tested. LGTM.

**Review by pankajastro — APPROVED** [review]

**Pull request merged**
